crates/mcp-server/src/lib.rs

```diff
@@ -43,17 +50,15 @@ where
 
     fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
         let mut this = self.project();
-        let mut buf = Vec::new();
-        // Default BufReader capacity is 8 * 1024, increase this to 2MB to the file size limit
-        // allows the buffer to have the capacity to read very large calls
-        let mut reader = BufReader::with_capacity(2 * 1024 * 1024, &mut this.reader);
 
-        let mut read_future = Box::pin(reader.read_until(b'\n', &mut buf));
+        let mut reader = this.reader.as_mut();
+        let mut read_future = Box::pin(reader.read_until(b'\n', this.buf));
         match read_future.as_mut().poll(cx) {
             Poll::Ready(Ok(0)) => Poll::Ready(None), // EOF
             Poll::Ready(Ok(_)) => {
                 // Convert to UTF-8 string
-                let line = match String::from_utf8(buf) {
+                // std::mem::take will clear the buf with its default value (Vec::empty)
+                let line = match String::from_utf8(std::mem::take(this.buf)) {
```

Comment:
discussed offline, its probably better to clear() after reading instead of mem:take
```
The this.buf inside the struct is replaced with an empty buffer of capacity 0, so next time around it may re-allocate from scratch when you push more data. This is fine for correctness, but it forfeits the advantage of a growable buffer that can be reused.
```